draft-pda-protocol.md

```diff
@@ -573,116 +609,281 @@ PA protocol. The structure of the leader share is similarly protocol specific:
 
-The response is structured as a sequence of *sub-responses*, where the i-th
-sub-response corresponds to the sub-request for each i. The structure of each
-sub-response is specific to the PA protocol:
+The response consists of the helper's updated state and a sequence of
+*sub-responses*, where the i-th sub-response corresponds to the sub-request for
+each i. The structure of each sub-response is specific to the PA protocol:
 
 ~~~
 struct {
-   PAVerifyResp seq<1..2^24-1>;
-} PAVerifyRespSeq;
+  opaque helper_state<0..2^16>;
+  PAAggregateSubResp seq<1..2^24-1>;
+} PAAggregateResp;
 
 struct {
-  PAProto;
-  select (PAVerifyResp.proto) {
-    case prio: PrioVerifyResp;
-    case hits: HitsVerifyResp;
+  PAProto proto;
+  select (PAAggregateSubResp.proto) {
+    case prio: PrioAggregateSubResp;
+    case hits: HitsAggregateSubResp;
   }
-} PAVerifyResp;
+} PAAggregateSubResp;
 ~~~
 
-For each sub-request `PAVerifyReq`, the helper computes the corresponding
-sub-response as follows. It first checks that checks that `PAVerifyReq.proto ==
-PAParam.proto`. If not, it aborts and alerts the leader with "incorrect protocol
-for sub-request". Otherwise, It computes the HPKE context as
+The helper handles each sub-request `PAAggregateSubReq` as follows.  It first
+checks that checks that `PAAggregateSubReq.proto == PAParam.proto`. If not, it
+aborts and alerts the leader with "incorrect protocol for sub-request".
+Otherwise, It computes the HPKE context as
 
 ~~~
-context = SetupBaseR(PAVerifyReq.enc, sk, [TODO])
+context = SetupBaseR(PAAggregateSubReq.enc, sk, [TODO])
 ~~~
 
-where `sk` is the secret key corresponding to the HPKE config. Next, it computes
-the body of the `PAVerifyResp` according to the PA protocol.
+where `sk` is its HPKE key and computes the body of the `PAAggregateSubResp`. It
+then updates its state according to the PA protocol. After processing all of the
+sub-requests, the helper encrypts its updated state and constructs its response
+to the aggregate request.
 
-[OPEN ISSUE: encrypt and store Helper state at the Leader]
+##### Helper State
 
-## Collect {#pa-collect}
+The helper state is an optional parameter of an aggregate request that the can
+helper use to carry state accross requests. At least part of the state will
+usually need to be encrypted in order to protect user privacy. However, the
+details of precisely how the state is encrypted and the information that it
+carries is up to the implementation of the helper.
+
+#### Output Share Request
+
+Once the aggregators have verified at least as many reports as required for the
+PA task, the leader issues an *output share request* to the helper. The helper
+responds to this request by extracting its output share from its state and
+encrypting it under the collector's HPKE public key.
+
+The leader sends a POST request to `[helper]/output_share` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+  opaque helper_state<0..2^16>;
+} PAOutputShareReq;
```

Comment:
Hmm, where does the batch window come into play here? Shouldn't it be the case that aggregate outputs are only produced for a given batch window once the size is hit? Do the collector's request parameters need to fold into this OutputShareReq?

Reply:
Ah, good catch. I added the `batch_start` and `batch_end` parameters so that the helper can enforce the `batch_window` and `batch_size` parameters.